lib/stripe/instrumentation.rb

```diff
@@ -2,23 +2,65 @@
 
 module Stripe
   class Instrumentation
-    class RequestEvent
+    # Event emitted on `request_begin` callback.
+    class RequestBeginEvent
+      attr_reader :method
+      attr_reader :path
+
+      # Arbitrary user-provided data in the form of a Ruby hash that's passed
+      # from subscribers on `request_begin` to subscribers on `request_end`.
+      # `request_begin` subscribers can set keys which will then be available
+      # in `request_end`.
+      #
+      # Note that all subscribers of `request_begin` share the same object, so
+      # they must be careful to set unique keys so as to not conflict with data
+      # set by other subscribers.
+      attr_reader :user_data
+
+      def initialize(method:, path:, user_data:)
+        @method = method
+        @path = path
+        @user_data = user_data
+        freeze
+      end
+    end
+
+    # Event emitted on `request_end` callback.
+    class RequestEndEvent
       attr_reader :duration
       attr_reader :http_status
       attr_reader :method
       attr_reader :num_retries
       attr_reader :path
 
-      def initialize(duration:, http_status:, method:, num_retries:, path:)
+      # Arbitrary user-provided data in the form of a Ruby hash that's passed
+      # from subscribers on `request_begin` to subscribers on `request_end`.
+      # `request_begin` subscribers can set keys which will then be available
+      # in `request_end`.
+      attr_reader :user_data
+
+      def initialize(duration:, http_status:, method:, num_retries:, path:,
+                     user_data: nil)
         @duration = duration
         @http_status = http_status
         @method = method
         @num_retries = num_retries
         @path = path
+        @user_data = user_data
         freeze
       end
     end
 
+    # This class was renamed for consistency. This alias is here for backwards
+    # compatibility.
+    RequestEvent = RequestEndEvent
+
+    # Returns true if there are a non-zero number of subscribers on the given
+    # topic, and false otherwise.
+    def self.any_subscribers?(topic)
```

Comment:
Added this so that we can short circuit early if topics don't have subscribers so that we don't have the additional overhead of instantiating objects like `RequestBeginEvent`.

If you think about it, probably the vast majority of stripe-ruby installations are not going to have any callbacks configured, so skipping the work if possible makes some amount of sense.